### ISSUE TITLE
Getting 'Bad State' exception when disconnecting a connection in mid auto-reconnect

### ISSUE THREAD

**Issue by kmiller15211** [shamblett/mqtt5_client#110]
I have a situation where I am creating a websocket IoT connection to AWS via Cognito credentials. I have auto reconnect, and resubscribeOnAutoReconnect set to true, keepalive set to 5 seconds. 

I have a scenario where, when I get new AWS Cognito credentials (due to the activity token expiring, or the wifi goes off and then back on again, etc.), I want to stop the old connection, and start a new one with the new AWS Cognito credentials. What is happening is that the old connection is in the background trying to autoreconnect. So, even though I call disconnect() on the old connection, when that connect call completes and it tries to run the code in mqtt_connection_handler_base.dart that follows, it gets the following exception: 

The exception: "Bad state: Cannot add new events after calling close"
The throwing source: ...mqtt5_client-4.5.0/lib/src/connectionhandling/mqtt_connection_handler_base.dart

Tracing through the timeline: 
* I'm connected with connection A, and everything is happy. 
* I turn wifi off and wait for a bit for the connection to go down.
   * Connection A detects the loss of keepalive heartbeats, it goes into it's autoreconnect state. 
* I turn wifi back on.
* My code detects the wifi state is on again, and regenerates cognito credentials - callbacks are called to notify clients of this. 
* The code that's managing the IoT connections gets this callback and reacts by trying to create a brand new connection:
   * It then calls disconnect() on connection A, and stops the listener of the connection A stream
   * It creates a brand new IoT connection, connection B, with new Cognito credentials and connects
   * Connection B connects, starts listening, etc. Everything with Connection B is happy.
* Eventually, Connection A's autoreconnect call (to connect()) times out, and then it goes to this else case in mqtt_connection_handler_base: 

```dart
    connectionStatus = await connect(server, port, connectionMessage);
    autoReconnectInProgress = false;
    if (connectionStatus.state == MqttConnectionState.connected) {
        [REDACTED]
    } else {
      MqttLogger.log(
          'MqttConnectionHandlerBase::autoReconnect - auto reconnect failed - re trying');
      clientEventBus!.fire(MqttAutoReconnect());
    }
```
It is here where I believe it's trying to fire on the clientEventBus, and since the connection has been disconnected/closed, it's going down. 

This results in an unhandled exception. It's not causing any major issues (like a tree falling in the forest where there's no one around to hear it), but I'd like to clean this up, if possible. 

Maybe this line: `clientEventBus!.fire(MqttAutoReconnect());` should use a "?" instead of a "!" so that it quietly fails instead of throws? Or maybe there's some other API that I'm not seeing that I can call to allow connection A to finish cleaning up before creating B (undesirable, as I don't want to block for seconds)? 

Thank you in advance. Sorry if this isn't following some GitHub etiquette, I haven't entered many issues. 

**TLDR Details:** 
Deploying/running on an Android Pixel 7
Building with Android Studio Koala 2024.1.1 Patch 1
Using the MQTT 5 library, version: mqtt5_client: ^4.5.0

**Comment by shamblett** [comment]
Sorry for the late response I've been on hols.

Your thinking is correct, what your seeing is a 'closure' effect in Dart, the code has closed over the old(connection A) instantiation of the event bus, which is closed but not yet garbage collected, hence the '!' check works. These can be tricky to fix but it does need doing so I will look at this.

Your issue is fine BTW, the more info the better if like me you are a package maintainer. You would be surprised at the amount of issues I get that just say things like 'I have set autoreconnect on the client but its not working', go figure.

**Comment by shamblett** [comment]
OK, I've added an update for this, I've also written a test for it which seems OK but this is a tricky area. The package has been re released at version 4.5.2, please could you re test this, thanks.

**Comment by kmiller15211** [comment]
After the update to 4.5.3, here is my experience:

To start this scenario, I had the IoT connection connected and everything happy. Then, I turned off the Wifi on my Pixel: 
* Connection 0 (the initial connection) detects the network going down and begins to autoconnect (onAutoReconnecting)
* (I wait a few seconds)
* I turn wifi back on
* My code detects that Wifi is back, calls disconnect on connection 0 and sets and creates Connection 1 (setting MQTT connection member variable to the new instance in the process). This uses the same settings/credentials as connection 0. I would assume that connection 0 is still retrying in the background, but it is never heard from again in my logs. 
* Connection 1 connects (onConnected)
* Connection 1 subscribes to topics and everything seems fine. 
* (Approximately 5 seconds later, connection 2 seems to have a spurious disconnect/reconnect)
* Everything seems to work from this point on, although the spurious disconnect/reconnect feels like a minor bug. 

In more detail, from my log. Every time I create a new IoT connection, I increment the number in the logger, so I can tell the two objects apart. Below, you'll see "IotCognitoConnection(X)", where X is the ID of the instance. This coorelates with "Connection 0" or "Connection 1" above. 

```
// I pull the wifi:
I/flutter (10088): [2024/09/20] [08:08:11.846] INFO - NetworkConnection: Network status change: [ConnectivityResult.none]
I/flutter (10088): [2024/09/20] [08:08:11.847] FINE - CognitoMgr: Cognito reacting to network status change: [ConnectivityResult.none]
I/flutter (10088): [2024/09/20] [08:08:26.511] FINE - IotCognitoConnection(0): onAutoReconnecting

// Wifi is restarted: 
I/flutter (10088): [2024/09/20] [08:08:28.547] INFO - NetworkConnection: Network status change: [ConnectivityResult.wifi]
...

// At this point, the system detects wifi is back, and it starts by disconnecting connection 0.
I/flutter (10088): [2024/09/20] [08:08:28.987] FINE - IotCognitoConnection(0): onDisconnected
I/flutter (10088): [2024/09/20] [08:08:28.987] INFO - IotCognitoConnection(0): Disconnected
...
// Next, it builds a new connection object - Connection 1.
I/flutter (10088): [2024/09/20] [08:08:29.012] FINE - IotCognitoConnection(1): Attempting IoT connection
...
I/flutter (10088): [2024/09/20] [08:08:29.354] FINE - IotCognitoConnection(1): onConnected
I/flutter (10088): [2024/09/20] [08:08:29.356] INFO - IotCognitoConnection(1): Connected!
I/flutter (10088): [2024/09/20] [08:08:29.357] FINE - IotCognitoConnection(1): Waiting for the topic: us-east-1:REDACTED to finish subscription
I/flutter (10088): [2024/09/20] [08:08:29.421] INFO - IotCognitoConnection(1): New subscription detected: us-east-1:REDACTED
...
I/flutter (10088): [2024/09/20] [08:08:29.859] INFO - IotCognitoConnection(1): Subscription to: us-east-1:REDACTED completed

// Here, about 5 seconds later, it seems like there's a spurious disconnection of connection 1 (perhaps caused by Connection 0 going down in the background?)
I/flutter (10088): [2024/09/20] [08:08:34.022] FINE - IotCognitoConnection(1): onAutoReconnecting
I/flutter (10088): [2024/09/20] [08:08:34.332] FINE - IotCognitoConnection(1): onConnected
I/flutter (10088): [2024/09/20] [08:08:34.341] FINE - IotCognitoConnection(1): onAutoReconnected
I/flutter (10088): [2024/09/20] [08:08:34.445] INFO - IotCognitoConnection(1): New subscription detected: us-east-1:REDACTED
I/flutter (10088): [2024/09/20] [08:08:39.207] FINE - IotCognitoConnection(1): onAutoReconnecting
I/flutter (10088): [2024/09/20] [08:08:39.520] FINE - IotCognitoConnection(1): onConnected
I/flutter (10088): [2024/09/20] [08:08:39.527] FINE - IotCognitoConnection(1): onAutoReconnected
I/flutter (10088): [2024/09/20] [08:08:39.608] INFO - IotCognitoConnection(1): New subscription detected: us-east-1:REDACTED
```

I hope this helps.

**Comment by shamblett** [comment]
OK so the bad state exception has gone as expected. I can tell what the reason for the auto reconnect after 5 seconds is from your log, I'll need a client log for this to go any further..
